refactor: preferences & programs data model merged

emilyjablonski opened the pull request:
# Pull Request Template

## Issue Overview

This PR addresses #2752

- [x] This change addresses only certain aspects of the issue

## Description

This PR updates the preferences and programs data model to be able to support the [Figma designs](https://www.figma.com/file/Y8bFd0aSz7eYJ2TCNid6s3/bloom-partners?node-id=5725%3A182890) for the new preferences and programs management. It also merges the preference and program data models since they were nearly identical, and since the new designs for managing them are nearly identical.

Overview of the refactor:
The preference and program entities have been merged into a new entity that encompasses both of them called `Multiselect Question`
MultiselectQuestion
* Previously, preferences had their titles and option titles stored as just keys that map into our translation files. To support the new design, these need to be just English strings.
* The multiselect question has a new field called `applicationSection` to distinguish what section of an application it belongs in, right now we just have `preferences` and `programs`.

MultiselectOption
* A preference set had its options stored in a `formMetadata` field - this is now a new object called `options`, and the root setting of `hideFromListing` has been moved up to the question level. We no longer support custom select text, and don't need `hideGenericDecline` due to our new design around adding opt out text.
* An individual option now supports links.
* The data types for what sorts of extra data an option can collect has been reduced to just an address.
* The exclusive flag if true indicates that when that option is selected, it can only be selected exclusively, so every other option must automatically uncheck. Our decline options behave this way.

## How Can This Be Tested/Reviewed?

There are four main areas where we use preferences and programs that need to be tested:
(1) Public Listing View
* Should display all preferences and their links/description unless the `hideFromListing` tag is true

(2) Public Application
* Should now have a new section in the application labelled `Programs` that is distinct from `Preferences` if the application does have programs.
* If every option in a multiselect question is exclusive, render as a radio question
* The application summary page before submission should render the answers as expected

(3) Partners Application
* Data for preferences and programs when submitted should be reflected accurately
* The paper application form on partners should submit accurately
* The CSV export should display the data for preferences and programs accurately

(4) Partners Listing Form
* Jurisdictional preferences and programs should be appropriately fetched when selected them on a listing
* They should both save accurately

## Checklist:

- [x] My code follows the style guidelines of this project
- [ ] I have added QA notes to the issue with applicable URLs
- [x] I have performed a self-review of my own code
- [x] I have reviewed the changes in a desktop view
- [x] I have reviewed the changes in a mobile view
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have assigned reviewers
- [ ] I have run `yarn generate:client` and/or created a migration if I made backend changes that require them
- [ ] I have exported any new pieces added to ui-components
- [ ] My commit message(s) is/are polished, and any breaking changes are indicated in the message and are well-described
- [ ] Commits made across packages purposefully have the same commit message/version change, else are separated into different commits

## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.

emilyjablonski commented:
Remaining pieces: CSV export, Cypress tests

emilyjablonski commented:
Back to WIP, discovering some weird things w application endpoints

seanmalbert commented on backend/core/src/migration/1658871879940-multiselect-question.ts:
```diff
@@ -0,0 +1,287 @@
+import { MigrationInterface, QueryRunner } from "typeorm"
+// importing these causes issue with build command compiling to dist, because it changes the expected structure, so we need to find an alternative to providing these
```
@emilyjablonski and @YazeedLoonat

seanmalbert commented on backend/core/src/translations/services/translations.service.ts:
```diff
@@ -95,10 +95,26 @@ export class TranslationsService extends AbstractServiceFactory<
+        console.log(
+          "pushing",
+          `listingMultiselectQuestions[${i}].multiselectQuestion.options[${j}].text`
+        )
```
remaining log

emilyjablonski commented on backend/core/src/listings/listings.service.ts:
```diff
@@ -148,9 +148,9 @@ export class ListingsService {
-      .orderBy({
-        "listingPreferences.ordinal": "ASC",
-      })
+      // .orderBy({
```
@seanmalbert Quick q for you here on why this would have been here before?

seanmalbert replied:
It ordered the preferences, so we could keep it for listingMultiselectQuestions unless we're ensuring the order somewhere else that's more efficient.

emilyjablonski replied:
We do sort elsewhere, and now we'd have to sort across application sections here, so I'll move forward by removing it from here 👍

emilyjablonski commented on backend/core/src/seeder/seeds/applications.ts:
```diff
@@ -13,53 +13,20 @@ import { Application } from "../../applications/entities/application.entity"
-      street2: "street2",
-      zipCode: "zip code",
+const getApplicationCreateDtoTemplate = (
+  jurisdictionString: string
```
The diff is odd here, the only difference is the added `jurisdictionString` prop

emilyjablonski commented on backend/core/test/authz/authz.e2e-spec.ts:
```diff
@@ -31,7 +31,7 @@ describe("Authz", () => {
   let adminAccessToken: string
   let userProfile: UserDto
-  const adminOnlyEndpoints = ["/preferences", "/units", "/translations"]
+  const adminOnlyEndpoints = ["/units", "/translations"]
```
We had preferences here, but not programs - thoughts on what the permissions should be for all multiselect questions? @seanmalbert

seanmalbert replied:
Read should be fully open and the rest of the perms should be locked to admin and jurisdictional admins (I think we want JAs to be able to create, edit and delete questions).

emilyjablonski commented on shared-helpers/src/multiselectQuestions.tsx:
```diff
@@ -0,0 +1,400 @@
+        return acc
+      }, {})
+
+      questionsFormData["application"][applicationSection][question.key] = options
```
The suggestion they give here is to use a map which feels odd here, looking for other suggestions if y'all have any!

seanmalbert commented on backend/core/src/translations/services/translations.service.ts:
```diff
@@ -95,10 +95,22 @@ export class TranslationsService extends AbstractServiceFactory<
-      pathsToFilter.push(`listingPreferences[${i}].preference.description`)
-      pathsToFilter.push(`listingPreferences[${i}].preference.subtitle`)
+    for (let i = 0; i < listing.listingMultiselectQuestions.length; i++) {
+      pathsToFilter.push(`listingMultiselectQuestions[${i}].multiselectQuestion.text`)
```
Should we also translate link titles?

seanmalbert commented on sites/partners/src/applications/PaperApplicationForm/FormTypes.ts:
```diff
@@ -57,8 +56,8 @@ export type ApplicationTypes = {
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  preferences: Record<string, any>
-  programs: Record<string, string>
+  preferences?: Record<string, any>
+  programs?: Record<string, string>
```
should'n't these both be the same type now and can we explicitly type it?

emilyjablonski replied:
They should definitely be the same! This is representing the react-hook-form structure, which is super wonky and difference between checkboxes and radios, so I'll update the type but it will still be pretty open.

seanmalbert commented on sites/partners/src/applications/PaperApplicationForm/sections/FormMultiselectQuestions.tsx:
```diff
@@ -0,0 +1,147 @@
+  }, [questions, applicationSection])
+
+  if (questions?.length === 0) {
+    return <></>
```
is this different than returning `null`?

emilyjablonski replied:
I don't believe so and null looks to be preferable in the docs, I'll update 👍

seanmalbert commented on sites/public/pages/applications/review/demographics.tsx:
```diff
@@ -24,15 +24,20 @@ import {
   const { conductor, application, listing } = useFormConductor("demographics")
-  const currentPageSection = 5
+  let currentPageSection = 4
+  if (listingSectionQuestions(listing, ApplicationSection.programs)?.length) currentPageSection += 1
+  if (listingSectionQuestions(listing, ApplicationSection.preferences)?.length)
+    currentPageSection += 1
```
Is it possible to have this kind of logic in the ApplicationConductor, so we don't have to have it in multiple files?

emilyjablonski replied:
I would love to, it seems like there are a couple places we could choose to do that in ApplicationConductor, would you be okay splitting that into a separate ticket?

seanmalbert replied:
Sounds good, @emilyjablonski. Here's the issue https://github.com/bloom-housing/bloom/issues/2972.

seanmalbert reviewed (commented):
Whew, nice work, @emilyjablonski! I feel fatigued just from reviewing all this. I've got some questions I added, which may or may not require updates.

emilyjablonski commented:
@YazeedLoonat When I reseed on dev and just run the migration on this branch it fails, would you mind taking a look?

ludtkemorgan commented on sites/partners/src/applications/PaperApplicationDetails/sections/DetailsMultiselectQuestions.tsx:
```diff
@@ -1,54 +1,47 @@
+            <ViewItem label={listingQuestion?.multiselectQuestion.text}>
               {(() => {
-                if (!optionDetails?.claimed) return t("t.none")
+                const appQuestion = questions?.filter(
+                  (question) => question.key === listingQuestion?.multiselectQuestion.text
+                )[0]
```
Since you are just getting the first one that is returned you can do 
`questions?.find` instead

ludtkemorgan reviewed (approved):
No major issues spotted. Since this is such a big PR I'd recommend getting a third approver though before merging

emilyjablonski commented:
Bugs fixed in latest commit: (1) optional preferences no longer say `please select one`, fixed issue with saving the preferences panel when there is no jurisdiction selected yet (also removed the button from this state, since it doesn't make sense)

ColinBuyck reviewed (approved):
LGTM! Great work!

emilyjablonski commented:
good to go when https://github.com/bloom-housing/bloom/pull/2975 is resolved!
